src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+# Implementation of System.Text.RegularExpressions
+
+The implementation uses a typical NFA approach that supports back references. Patterns are parsed into a tree (`RegexTree`), translated into an abstract tree (`RegexCode`) by a writer (`RegexWriter`), and then either used in an interpreter (`RegexInterpreter`) or compiled to IL which is executed (`CompiledRegexRunner`). Both of these are instances of `RegexRunner`: in the case of the compiled runner, one must generate a new one from the `RegexCode` using a factory each time the pattern is to be executed.
```

Comment:
> translated into an abstract tree

RegexCode isn't a tree.  It's a sequence of instructions, basically the regex engine's IR or bytecode.

> Both of these are instances

Sounds better to my ear as "both of these are derived from `RegexRunner`

> in the case of the compiled runner, one must generate a new one from the `RegexCode` using a factory each time the pattern is to be executed

Not quite.  Regardless of whether the runner is an interpreted or compiled runner, an instance of the runner is cached on the Regex instance.  As long as the Regex instance is used only by one thread at a time, every use will end up using that same cached runner instance.  If it's being used concurrently such that one thread has taken the cached runner instance, then the other one or more threads trying to use it concurrently will be forced to create their own runner instance: in the case of a compiled instance, it'll ask the compiled instance's factory to instantiate a new runner, otherwise we instantiate the interpreter.

Reply:
Yup, thanks all of these things I figured out but hadn't updated the text.